staging/src/k8s.io/apiserver/pkg/util/flowcontrol/borrowing_test.go

```diff
@@ -0,0 +1,250 @@
+			}
+			clientset := clientsetfake.NewSimpleClientset(cfgObjs...)
+			informerFactory := informers.NewSharedInformerFactory(clientset, time.Second)
+			flowcontrolClient := clientset.FlowcontrolV1beta3()
+			clk := eventclock.Real{}
+			controller := newTestableController(TestableConfig{
+				Name:                   "Controller",
+				Clock:                  clk,
+				AsFieldManager:         ConfigConsumerAsFieldManager,
+				FoundToDangling:        func(found bool) bool { return !found },
+				InformerFactory:        informerFactory,
+				FlowcontrolClient:      flowcontrolClient,
+				ServerConcurrencyLimit: 24,
+				RequestWaitLimit:       time.Minute,
+				ReqsGaugeVec:           metrics.PriorityLevelConcurrencyGaugeVec,
+				ExecSeatsGaugeVec:      metrics.PriorityLevelExecutionSeatsGaugeVec,
+				QueueSetFactory:        fqs.NewQueueSetFactory(clk),
+			})
+
+			ctx, cancel := context.WithTimeout(context.Background(), 50*time.Second)
+			stopCh := ctx.Done()
+			controllerCompletionCh := make(chan error)
+
+			informerFactory.Start(stopCh)
+
+			status := informerFactory.WaitForCacheSync(ctx.Done())
+			if names := unsynced(status); len(names) > 0 {
+				t.Fatalf("WaitForCacheSync did not successfully complete, resources=%#v", names)
+			}
+
+			go func() {
+				controllerCompletionCh <- controller.Run(stopCh)
+			}()
+
+			// ensure that the controller has run its first loop.
+			err := wait.PollImmediate(100*time.Millisecond, 5*time.Second, func() (done bool, err error) {
+				return controller.hasPriorityLevelState(plcObjs[0].Name), nil
+			})
+			if err != nil {
+				t.Errorf("expected the controller to reconcile the priority level configuration object: %s, error: %s", plcObjs[0].Name, err)
+			}
+
+			concIntegrators := make([]fq.Integrator, 2)
+			reqInfo := &request.RequestInfo{
+				IsResourceRequest: false,
+				Path:              "/foobar",
+				Verb:              "GET",
+			}
+			noteFn := func(fs *flowcontrol.FlowSchema, plc *flowcontrol.PriorityLevelConfiguration, fd string) {}
+			workEstr := func() fcrequest.WorkEstimate { return fcrequest.WorkEstimate{InitialSeats: 1} }
+			qnf := fq.QueueNoteFn(func(bool) {})
+			var wg sync.WaitGroup
+			wg.Add(30)
+			// Launch 20 client threads for each flow
+			for flow := 0; flow < 2; flow++ {
+				username := usernames[flow]
+				flowUser := testUser{name: username}
+				rd := RequestDigest{
+					RequestInfo: reqInfo,
+					User:        flowUser,
+				}
+				concIntegrator := fq.NewNamedIntegrator(clk, username)
+				concIntegrators[flow] = concIntegrator
+				exec := func() {
+					concIntegrator.Inc()
+					clk.Sleep(250 * time.Millisecond)
+					concIntegrator.Dec()
+				}
+				nThreads := 24 - flow*18
+				for thread := 0; thread < nThreads; thread++ {
+					go func() {
+						wg.Done()
+						wait.Until(func() { controller.Handle(ctx, rd, noteFn, workEstr, qnf, exec) }, 0, ctx.Done())
+					}()
+				}
+			}
+			wg.Wait()
```

Comment:
I was expecting to see this at the end or via defer, I thought the threads would run until the `cancel()` on line 221 is executed?

Reply:
Wait, the wg is to make sure the thread has started, not to make sure it has ended?

Reply:
Considering the Sleep(10) and the fact that everything passed prior to adding this call, I'm not sure this waitgroup is having an effect on the test.

Reply:
Yes, that wait group is to make sure the goroutines have started.